zst/trailer.go

```diff
@@ -1,159 +1,37 @@
+	FileType = "zst"
+	Version  = 2
 )
 
-// XXX we should make generic trailer package and share between microindex and zst
-
-type Trailer struct {
-	Length        int
-	Magic         string
-	Version       int
+type FileMeta struct {
 	SkewThresh    int
 	SegmentThresh int
-	Sections      []int64
-}
-
-var ErrNotZst = errors.New("not a zst object")
-
-func newTrailerRecord(zctx *zed.Context, skewThresh, segmentThresh int, sections []int64) (*zed.Value, error) {
-	sectionsType := zctx.LookupTypeArray(zed.TypeInt64)
-	cols := []zed.Column{
-		{MagicField, zed.TypeString},
-		{VersionField, zed.TypeInt32},
-		{SkewThreshField, zed.TypeInt32},
-		{SegmentThreshField, zed.TypeInt32},
-		{SectionsField, sectionsType},
-	}
-	typ, err := zctx.LookupTypeRecord(cols)
-	if err != nil {
-		return nil, err
-	}
-	builder := zed.NewBuilder(typ)
-	return builder.Build(
-		zed.EncodeString(MagicVal),
-		zed.EncodeInt(VersionVal),
-		zed.EncodeInt(int64(skewThresh)),
-		zed.EncodeInt(int64(segmentThresh)),
-		encodeSections(sections)), nil
-}
-
-func encodeSections(sections []int64) zcode.Bytes {
-	var b zcode.Builder
-	for _, s := range sections {
-		b.Append(zed.EncodeInt(s))
-	}
-	return b.Bytes()
-}
-
-func readTrailer(r io.ReadSeeker, n int64) (*Trailer, error) {
-	if n > TrailerMaxSize {
-		n = TrailerMaxSize
-	}
-	if _, err := r.Seek(-n, io.SeekEnd); err != nil {
-		return nil, err
-	}
-	buf := make([]byte, n)
-	cc, err := r.Read(buf)
-	if err != nil {
-		return nil, err
-	}
-	if int64(cc) != n {
-		// This shouldn't happen but maybe could occur under a corner case
-		// or I/O problems.
-		return nil, fmt.Errorf("couldn't read trailer: expected %d bytes but read %d", n, cc)
-	}
-	stream, err := zngio.FindTrailer(buf)
-	if err != nil {
-		return nil, err
-	}
-	rec, _ := zngio.NewReader(bytes.NewReader(stream), zed.NewContext()).Read()
-	if rec == nil {
-		return nil, errors.New("zst trailer not found")
-	}
-	//XXX This should be simplified and made more robust.  See issue #3476.
-	trailer, _ := recordToTrailer(rec)
-	if trailer != nil {
-		trailer.Length = len(stream)
-		return trailer, nil
-	}
-	return nil, errors.New("zst trailer not found")
 }
 
-func trailerVersion(rec *zed.Value) (int, error) {
-	version, err := rec.AccessInt(VersionField)
+func readTrailer(r io.ReaderAt, n int64) (*FileMeta, []int64, error) {
+	trailer, err := zngio.ReadTrailer(r, n)
 	if err != nil {
-		return -1, errors.New("zst version field is not a valid int32")
-	}
-	if version != VersionVal {
-		return -1, fmt.Errorf("zst version %d found while expecting version %d", version, VersionVal)
+		return nil, nil, err
 	}
-	return int(version), nil
-}
-
-func recordToTrailer(rec *zed.Value) (*Trailer, error) {
-	var trailer Trailer
-	var err error
-	trailer.Magic, err = rec.AccessString(MagicField)
-	if err != nil || trailer.Magic != MagicVal {
-		return nil, ErrNotZst
+	if trailer.Type != FileType {
+		return nil, nil, fmt.Errorf("not a zst file: trailer type is %q", trailer.Type)
```

Comment:
While you're here, maybe change zst to ZST in all these error strings?

Reply:
This is being changed on zst PR